### PR TITLE
Make error codes into strings

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of #67061.

I used this script to convert all error codes into strings:

<details>

```python
from os import listdir
from os.path import isdir, isfile, join


def remove_import_and_replace_err_codes(f_path):
    result = []
    with open(f_path) as f:
        lines = f.read().split("\n")
        i = 0
        found = False
        while i < len(lines):
            if lines[i] == 'use rustc_error_codes::*;':
                found = True
                pass
            elif ' E0' in lines[i] and not lines[i].strip().startswith('//') and not lines[i].endswith('\\') and not '"' in lines[i]:
                found = False
                parts = lines[i].split(' E0')
                p = parts[1].split(',', 1)
                if len(p) < 2:
                    result.append(lines[i])
                else:
                    result.append('{} "E0{}",{}'.format(parts[0], p[0], p[1]))
            elif '(E0' in lines[i] and not lines[i].strip().startswith('//') and not lines[i].endswith('
\\') and not '"' in lines[i]:
                found = False
                parts = lines[i].split('(E0')
                p = parts[1].split(')', 1)
                p2 = parts[1].split(',', 1)
                if len(p2[0]) < len(p[0]) and len(p2) > 1:
                    p = p2
                if len(p) < 2:
                    result.append(lines[i])
                else:
                    result.append('{}("E0{}"){}'.format(parts[0], p[0], p[1]))
            else:
                if len(lines[i]) != 0 or found is False:
                    result.append(lines[i])
                found = False
            i += 1
    with open(f_path, 'w') as f:
        f.write('\n'.join(result))


def remove_dependency(f_path):
    result = []
    with open(f_path) as f:
        lines = f.read().split("\n")
        i = 0
        while i < len(lines):
            if lines[i].startswith('rustc_error_codes = '):
      	        pass
            else:
                result.append(lines[i])
            i += 1
    with open(f_path, 'w') as f:
        f.write('\n'.join(result))


def loop_dirs(cur_dir):
    for entry in listdir(cur_dir):
        f = join(cur_dir, entry)
        if isfile(f):
            if entry.endswith(".rs"):
                remove_import_and_replace_err_codes(f)
            elif entry.endswith("Cargo.toml"):
                remove_dependency(f)
        elif isdir(f) and not entry.startswith("librustc_error_codes") and f.startswith('src/lib'):
            loop_dirs(f)


loop_dirs('src')
```
</details>

r? @Centril

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
I am opposed to just using strings here -- that seems really error prone, and also harder to search for in tidy and the like. I would prefer that we add a `ErrorCode(pub usize)` struct to librustc_errors and use that everywhere (i.e., `ErrorCode(303)`). We should strive to avoid strings in any location where there is a more canonical type, or just not do this at all for now.

**Comment by GuillaumeGomez** [comment]
@Mark-Simulacrum It comes after a discussion with @eddyb and @Centril. I guess I'll let them explain in details **why**. ;)

**Comment by eddyb** [comment]
I am opposed to doing anything here unless explicitly sanctioned by @rust-lang/wg-diagnostics and I apologize if there was any confusion about this on Discord.

**Comment by estebank** [comment]
Even if we wanted to stringify these (and I don't think we do), the diagnostic macros could accept the code without having to surround it with quotes.

File: src/libsyntax/diagnostics/macros.rs [hunk 1]
```diff
@@ -1,112 +1,112 @@
...
 #[macro_export]
 macro_rules! span_fatal {
-    ($session:expr, $span:expr, $code:ident, $($message:tt)*) => ({
+    ($session:expr, $span:expr, $code:expr, $($message:tt)*) => ({
         $crate::diagnostic_used!($code);
         $session.span_fatal_with_code(
             $span,
             &format!($($message)*),
-            $crate::errors::DiagnosticId::Error(stringify!($code).to_owned()),
+            $crate::errors::DiagnosticId::Error($code.to_owned()),
```
Comment (Centril) [inline src/libsyntax/diagnostics/macros.rs]: This seems like an unnecessary change; we can keep the identifiers and continue using `stringify!($code)`. The only thing that's important from a recompilation POV is that `diagnostic_used!(...)` be removed or that if it exists for the purposes of tidy, that it shouldn't do anything (i.e. have an empty expansion).

**Comment by Centril** [comment]
The only thing that's important to me here is that we remove the dependency of `librustc_error_codes` from librustc_parser and so on. Otherwise you have to recompile everything every time you want to add a new error code. (I guess that doesn't matter if we have frozen new error code addition.)

**Comment by Centril** [comment]
> I would prefer that we add a `ErrorCode(pub usize)` struct to librustc_errors and use that everywhere (i.e., `ErrorCode(303)`). We should strive to avoid strings in any location where there is a more canonical type, or just not do this at all for now.

That would also be fine by me since it means that we can still remove the `librustc_error_codes` dependency from librustc_parser et al.

**Comment by GuillaumeGomez** [comment]
We're evolving towards strings identifiers like "[LifetimeError.Mismatch]". Therefore, strings make more sense to prepare for this switch.

**Comment by Centril** [comment]
@GuillaumeGomez Well there seems to be a lack of consensus in favor of strings, so could we limit this PR for now to just the tidy check and removing references to `librustc_error_codes` in the various `Cargo.toml`s in this PR? We can always move to strings later.

**Comment by GuillaumeGomez** [comment]
Euh we can't really. Currently it's using consts so we have to pick something in order to remove the dependency.

File: src/libsyntax/diagnostics/macros.rs [hunk 2]
```diff
@@ -1,112 +1,112 @@
 #[macro_export]
 macro_rules! diagnostic_used {
-    ($code:ident) => (
+    ($code:expr) => (
         let _ = $code;
```
Comment (Centril) [inline src/libsyntax/diagnostics/macros.rs]: @GuillaumeGomez I don't think this macro has any function at all before or after the PR. These constants are all marked as `pub` inside `librustc_error_codes` so they will not be deemed unused if `diagnostic_used!(...);` is missing. Therefore the macro can be removed wholesale and any calls to it. Also, all you've done here is to replace `stringify!($some_ident)` with string literals instead. Keeping the identifiers and continuing to use `stringify!(...)` should work just as well.

**Comment by JohnTitor** [comment]
Ping from triage: @GuillaumeGomez this PR is idle for a month, could you address Centril's review?

**Comment by GuillaumeGomez** [comment]
I need to update it, indeed!

**Comment by GuillaumeGomez** [comment]
Updated: however one issue remains: since error codes are now strings, the JSON rendering changes. How can I implement `RustcEncodable` on `librustc_errors::DiagnosticCode` by hand so the `code` field doesn't get handled like a string? I didn't find any "manual" implementation of it...

**Comment by Centril** [comment]
I fixed the dependency issue in https://github.com/rust-lang/rust/pull/68353. Beyond that, it's not really clear to me what benefits the stringification brings, and there seems to be a lack of consensus in favor of that change for now. So I'll r? @Mark-Simulacrum @oli-obk @estebank.

**Comment by eddyb** [comment]
@GuillaumeGomez That's because the trait is named `Encodable`.

**Comment by GuillaumeGomez** [comment]
@Centril It appeared that error codes are used by people. So a solution to improve their usability was to allow to have "error codes" looking like "lifetime-mismatch" instead of "E0XXX". If we handle error codes with strings instead of enums, we can then use them however we want.

**Comment by Centril** [comment]
@GuillaumeGomez I mean, it could just as well look like `lifetime_mismatch` (still `:ident`), or `lifetime::mismatch` (`:path`).

**Comment by GuillaumeGomez** [comment]
Rereading from the conversation, I'm not even sure that @rust-lang/wg-diagnostics came to a conclusion on this. So let's wait to hear from them?

Or maybe you had some news on your side @eddyb ?

**Comment by eddyb** [comment]
> Or maybe you had some news on your side @eddyb ?


I am not involved in anything where this came up since our Discord discussion, so no.

**Comment by Centril** [comment]
Note that @rust-lang/wg-diagnostics is not really a group that holds regular meetings, has a well-defined membership, and whatnot. For now it's just @oli-obk, @estebank, @davidtwco, and myself.

**Comment by oli-obk** [comment]
I think moving away from numeric error codes is a good thing, I do worry about the checks for duplicate use and undocumented errors though. It seems the current impl is very hardcoded to things starting with `"E0` without the error emitters checking that you actually name your error codes this way

**Comment by Mark-Simulacrum** [comment]
@Centril noted elsewhere (not sure where) that the dependency breaking part of this has already been done. It is not clear to me that this PR makes things better wrt to stringification, as a result of that. I think my earlier comment (https://github.com/rust-lang/rust/pull/67086#issuecomment-562571752) still stands.

Splitting out the unused/definitely used checking code into a separate PR (or making that the only change in this PR) seems good, in theory, though I have not reviewed whether:
 * we need that code
 * it is (in fact) correct

**Comment by GuillaumeGomez** [comment]
Ok, let's wait and see then!

**Comment by GuillaumeGomez** [comment]
Got the comments after commenting...

@Mark-Simulacrum I can send the check for the double error codes usage in another PR if you want?

**Comment by Mark-Simulacrum** [comment]
Sure, that'd be great.

**Comment by GuillaumeGomez** [comment]
@oli-obk We can store the error codes inside an hashmap and make this check when throwing one, and then we could use this hashmap as reference for the possible duplicates? (even though I'm not completely sure how to correctly handle it, i.e. differentiating "normal" strings and error codes)

**Comment by GuillaumeGomez** [comment]
> Sure, that'd be great.

@Mark-Simulacrum Ok, sending the PR this evening or tomorrow.

**Comment by GuillaumeGomez** [comment]
ping @oli-obk

**Comment by oli-obk** [comment]
Reading the discussion on zulip (https://rust-lang.zulipchat.com/#narrow/stream/147480-t-compiler.2Fwg-diagnostics/topic/get.20rid.20of.20error.20codes/near/187901869) I don't think we're ready to do this right now. What we definitely don't want is to have quoted strings in the macro, since any non-numeric names will just be snake case names and thus valid idents that can be stringified internally.

So basically

* this PR should not be touching any call-sites of the error creation macros
* in the future we may want to accept error codes that are not in the `E0XXX` format
* we may want to have a translation step in the macros allowing us to use the numeric and named error codes interchangeably

File: Cargo.lock [hunk 3]
```diff
@@ -3855,7 +3855,10 @@ dependencies = [
  "rustc_ast_pretty",
  "rustc_attr",
  "rustc_data_structures",
+<<<<<<< HEAD
```
Comment (mati865) [inline Cargo.lock]: Bad merge

**Comment by joelpalmer** [comment]
Triaged
